clients/src/main/java/org/apache/kafka/common/utils/Bytes.java

```diff
@@ -0,0 +1,415 @@
+    public final static Comparator<byte[]> BYTES_LEXICO_COMPARATOR = new LexicographicByteArrayComparator();
+
+    public interface ByteArrayComparator extends Comparator<byte[]> {
+
+        int compare(final byte[] buffer1, int offset1, int length1,
+                    final byte[] buffer2, int offset2, int length2);
+    }
+
+    public static class LexicographicByteArrayComparator implements ByteArrayComparator {
+
+        @Override
+        public int compare(byte[] buffer1, byte[] buffer2) {
+            return compare(buffer1, 0, buffer1.length, buffer2, 0, buffer2.length);
+        }
+
+        public int compare(final byte[] buffer1, int offset1, int length1,
+                           final byte[] buffer2, int offset2, int length2) {
+
+            // short circuit equal case
+            if (buffer1 == buffer2 &&
+                    offset1 == offset2 &&
+                    length1 == length2) {
+                return 0;
+            }
+
+            // bring WritableComparator code local
+            int end1 = offset1 + length1;
+            int end2 = offset2 + length2;
+            for (int i = offset1, j = offset2; i < end1 && j < end2; i++, j++) {
+                int a = (buffer1[i] & 0xff);
+                int b = (buffer2[j] & 0xff);
+                if (a != b) {
+                    return a - b;
+                }
+            }
+            return length1 - length2;
+        }
+    }
+
+    /**
+     * Lexicographically compare two arrays.
+     *
+     * @param buffer1 left operand
+     * @param buffer2 right operand
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static int compareTo(final byte[] buffer1, final byte[] buffer2) {
+
+        return ((ByteArrayComparator) BYTES_LEXICO_COMPARATOR).compare(buffer1, 0, buffer1.length, buffer2, 0, buffer2.length);
+    }
+
+    /**
+     * Lexicographically compare two arrays.
+     *
+     * @param buffer1 left operand
+     * @param buffer2 right operand
+     * @param offset1 Where to start comparing in the left buffer
+     * @param offset2 Where to start comparing in the right buffer
+     * @param length1 How much to compare from the left buffer
+     * @param length2 How much to compare from the right buffer
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static int compareTo(final byte[] buffer1, final int offset1, final int length1,
+                                final byte[] buffer2, final int offset2, final int length2) {
+
+        return ((ByteArrayComparator) BYTES_LEXICO_COMPARATOR).compare(buffer1, offset1, length1, buffer2, offset2, length2);
+    }
+
+    /**
+     * Check if two byte arrays are lexicographically equal to each other.
+     *
+     * @param buffer1 left operand
+     * @param buffer2 right operand
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static boolean equals(final byte[] buffer1, final byte[] buffer2) {
```

Comment:
How is this different to `java.util.Arrays.equals`? The implementation will perform quite a bit worse, so do we actually need it?